docs/proposals/templates.md

```diff
@@ -0,0 +1,550 @@
+  requirement.
+* Template metadata.  Templates should be able to include metadata describing their purpose or links to further documentation and 
+  versioning information.
+
+
+## Proposed Implementation
+
+### Overview
+
+We began by looking at the List object which allows a user to easily group a set of objects together for easy creation via a 
+single CLI invocation.  It also provides a portable format which requires only a single file to represent an application.
+
+From that starting point, we propose a Template API object which can encapsulate the definition of all components of an 
+application to be created.  The application definition is encapsulated in the form of an array of API objects (identical to 
+List), plus a parameterization section.  Components reference the parameter by name and the value of the parameter is 
+substituted during a processing step, prior to submitting each component to the appropriate API endpoint for creation.
+
+The primary capability provided is that parameter values can easily be shared between components, such as a database password 
+that is provided by the user once, but then attached as an environment variable to both a database pod and a web frontend pod.
+
+In addition, the template can be repeatedly instantiated for a consistent application deployment experience in different 
+namespaces or Kubernetes clusters.
+
+Lastly, we propose the Template API object include a “Labels” section in which the template author can define a set of labels 
+to be applied to all objects created from the template.  This will give the template deployer an easy way to manage all the 
+components created from a given template.  These labels will also be applied to selectors defined by Objects within the template,
+allowing a combination of templates and labels to be used to scope resources within a namespace.  That is, a given template
+can be instantiated multiple times within the same namespace, as long as a different label value is used each for each
+instantiation.  The resulting objects will be independent from a replica/load-balancing perspective.
+
+Generation of parameter values for fields such as Secrets will be delegated to an [admission controller/initializer/finalizer](https://github.com/kubernetes/kubernetes/issues/3585) rather than being solved by the template processor.
+
+### API Objects
+
+**Template Object**
+```
+// Template contains the inputs needed to produce a Config.
+type Template struct {
+    unversioned.TypeMeta
+    kapi.ObjectMeta
+
+    // Optional: Parameters is an array of Parameters used during the
+    // Template to Config transformation.
+    Parameters []Parameter
+
+    // Required: A list of resources to create
+    Objects []runtime.Object
+
+    // Optional: ObjectLabels is a set of labels that are applied to every
+    // object during the Template to Config transformation
+    // These labels are also be applied to selectors defined by objects in the template
```

Comment:
this is the point/question I was trying to bring up during the in person discussion last week.  It seems to me that doing this requires a certain level of api/schema awareness on the part of the processor to understand "the object i'm processing is a ReplicationController, therefore i need to apply this label to the selector field" and whatever other very specific object types/fields where we'd want to do this.

In the existing implementation labels are only applied to the object's metadata, which is of course common across all objects, so it's a much simpler problem.

There's also the question of whether we'd really want all the labels applied to all the selectors.  Since all the pods created by the template are going to have the same labels (or at least a subset of matching labels) as defined by the Template, using those same labels for the selector is going to result in a selector that selects all the pods created by the template.  Likely that's not what is desired.

Perhaps I don't fully grasp what you were intending when you proposed that labels be applied to selector definitions.  Can you give a quick example of how it would work for a template that defines multiple replication controllers/pod templates?  (eg an app pod and a db pod)